### ISSUE TITLE
'Extension not installed on grain' best practices

### ISSUE THREAD

**Issue by ADringer** [dotnet/orleans#7833]
Hi,

Been running our Orleans application for a bit now and sometimes see the error `GrainExtensionNotInstalledException: Extension not installed on grain ...` in  the logs.

I know that this is due to a grain subscribing to a stream initially but upon the next activation it doesn't resume the subscription and throws the exception when a message arrives.

I'm wondering what the best practice/approach to this error is? In my application grains subscribe to different streams ad-hoc and we persist the id to the grains state for resuming, but something must have gone wrong somewhere as in some instances we are missing the persisted id.

Some questions/points:

1. We can do fire and forget delivery but I think this is glossing over our actual issue at this point.
2. Is there a way of capturing the subscribing grain and the stream id of the one failing? Then I could potentially add it to persistence to fix up the grain state.
3. If a producer sends out a message on a stream with say 5 subscribers and the thirds one throws the exception, do the last two still receive the message? I was thinking about possibly wrapping the producer bit in a try/catch so it can carry on it's logic, but not sure if all subscribers will receive the message.

Basically wondering what I can do to live grains to fix up their state. These are long life grains so do need to ideally correct them.

Thanks

**Comment by benjaminpetit** [comment]
Sorry I totally missed this issue.

You should implement the interface `IStreamSubscriptionObserver` in your grain, so you could unsubscribe from the stream when Orleans tries to attach a subscription to your activation.

For your point 3, yes it should.

I assume you are using SMS, correct?

**Comment by ADringer** [comment]
Hi, @benjaminpetit , thanks for getting back to me. 

Yes, I am using SMS.

I'm not quite following your suggestion for `IStreamSubscriptionObserver `. My issue is that a grain created a subscription a while ago, has deactivated, and then when reactivated it didn't resume the subscription so falls over when it recieves a message from the stream.

Does this method get called every time a message is sent to a subscription on the consuming grain? And if so, is resuming the subscription here an alternative to doing it on the `OnActivateAsync()` method?

**Comment by benjaminpetit** [comment]
The method from `IStreamSubscriptionObserver` is called everytime a message is sent to a subscription for the first time at the activation level.

So if your grain was subscribed to a stream, then is deactivated because of inactivity, then a new stream event arrives and trigger a new activation creation, the method will be called again.

**Comment by ADringer** [comment]
Okay cool, sounds like exactly what I need as I can resume the subscription then. 

At the moment I resume the streams I know about when the grain is activated - does it make any difference using the `IStreamSubscriptionObserver` instead to resume rather than doing them all upfront?

**Comment by benjaminpetit** [comment]
I think it's more elegant to use `IStreamSubscriptionObserver` and can handle your case "I don't want this subscription anymore, let's drop it". It also handles sbscription you may have missed for some reason. But both patterns are valid.
